src/nvim/ex_getln.c

```diff
@@ -2263,20 +2280,56 @@ static void draw_cmdline(int start, int len)
     msg_outtrans_len(ccline.cmdbuff + start, len);
 }
 
+void ui_ext_cmdline_char(int c, int shift)
+{
+    Array args = ARRAY_DICT_INIT;
+    ADD(args, STRING_OBJ(cstr_to_string((char *)(&c))));
+    ADD(args, INTEGER_OBJ(shift));
+    ui_event("cmdline_char", args);
+}
+
+void ui_ext_cmdline_show(void)
+{
+    Array args = ARRAY_DICT_INIT;
+    Array content = ARRAY_DICT_INIT;
+    Array text = ARRAY_DICT_INIT;
+    ADD(text, STRING_OBJ(cstr_to_string("Normal")));
+    ADD(text, STRING_OBJ(cstr_to_string((char *)(ccline.cmdbuff))));
+    ADD(content, ARRAY_OBJ(text));
+    ADD(args, ARRAY_OBJ(content));
+    ADD(args, INTEGER_OBJ(ccline.cmdpos));
+    if (ccline.cmdfirstc != NUL) {
+      ADD(args, STRING_OBJ(cstr_to_string((char *)(&ccline.cmdfirstc))));
```

Comment:
firstc is only : / ? etc which is defined by nvim. So there's no so many of them.